src/lib/protocols/credentials/__tests__/CredentialService.test.ts

```diff
@@ -0,0 +1,184 @@
+        ],
+      });
+    });
+
+    test('creates credential in OFFER_SENT state', async () => {
+      const credentialOffer = await credentialService.createCredentialOffer({
+        credDefId: 'Th7MpTaRZVRYnPiabds81Y:3:CL:17:TAG',
+        comment: 'some comment',
+        preview,
+      });
+      const [firstCredential] = await credentialService.getAll();
+
+      expect(firstCredential).toEqual(
+        expect.objectContaining({
+          createdAt: expect.any(Number),
+          id: expect.any(String),
+          offer: credentialOffer,
+          tags: {},
+          type: 'CredentialRecord',
+          state: CredentialState.OfferSent,
+        })
+      );
+    });
+
+    test(`emits stateChange event with ${CredentialState.OfferSent}`, async () => {
+      const eventListenerMock = jest.fn();
+      credentialService.on(EventType.StateChanged, eventListenerMock);
+
+      await credentialService.createCredentialOffer({
+        credDefId: 'Th7MpTaRZVRYnPiabds81Y:3:CL:17:TAG',
+        comment: 'some comment',
+        preview,
+      });
+
+      expect(eventListenerMock).toHaveBeenCalledWith(
+        expect.objectContaining({
+          newState: CredentialState.OfferSent,
+          credentialId: expect.any(String),
+        })
+      );
+    });
+  });
+
+  describe('processCredentialOffer', () => {
+    beforeEach(() => {
+      storageService = new StubStorageService();
+      credentialRepository = new Repository<CredentialRecord>(CredentialRecord, storageService);
+      credentialService = new CredentialService(wallet, credentialRepository);
+    });
+
+    test('creates credential in OFFER base on credential offer message', async () => {
+      const credentialOffer = await credentialService.createCredentialOffer({
+        credDefId: 'Th7MpTaRZVRYnPiabds81Y:3:CL:17:TAG',
+        comment: 'some comment',
+        preview,
+      });
+      const messageContext = new InboundMessageContext(credentialOffer);
+
+      await credentialService.processCredentialOffer(messageContext);
+      const [, secondCredential] = await credentialService.getAll();
+
+      expect(secondCredential).toEqual(
+        expect.objectContaining({
+          createdAt: expect.any(Number),
+          id: expect.any(String),
+          offer: credentialOffer,
+          tags: {},
+          type: 'CredentialRecord',
+          state: CredentialState.OfferReceived,
+        })
+      );
+    });
+
+    test(`emits stateChange event with ${CredentialState.OfferReceived}`, async () => {
+      const eventListenerMock = jest.fn();
+      credentialService.on(EventType.StateChanged, eventListenerMock);
+
+      const credentialOffer = await credentialService.createCredentialOffer({
+        credDefId: 'Th7MpTaRZVRYnPiabds81Y:3:CL:17:TAG',
+        comment: 'some comment',
+        preview,
+      });
+      const messageContext = new InboundMessageContext(credentialOffer);
+
+      await credentialService.processCredentialOffer(messageContext);
```

Comment:
I started with `acceptCredentialOffer` but then I realized I would be implementing more than it's necessary to pass the current test. We could also solve #60 `auto-accept` for connection in a similar way. We would have two methods `processInvitation` just to create a connection record and `acceptInvitation` to create `ConnectionRequestMessage`. The logic to decide if we want to `auto-accept` should be part of `Handler`.

Reply:
Agree!